omniscidb/QueryEngine/ResultSet.cpp

```diff
@@ -1498,17 +1504,65 @@ std::vector<std::pair<const int8_t*, size_t>> ResultSet::getChunkedColumnarBuffe
   std::vector<std::pair<const int8_t*, size_t>> retval;
   retval.reserve(1 + appended_storage_.size());
 
-  retval.emplace_back(
-      storage_->getUnderlyingBuffer() + storage_->getColOffInBytes(column_idx),
-      storage_->binSearchRowCount());
+  auto curr_storage_rows_to_fetch = storage_->binSearchRowCount();
+  size_t curr_storage_start_row = drop_first_;
+  size_t curr_storage_end_row =
+      keep_first_ != 0 ? keep_first_ : curr_storage_rows_to_fetch;
+  size_t total_handled_rows = 0;
+  // Should be cashed
+  size_t expected_row_count = rowCount();
+  // LOG(ERROR) << "storage";
+  // LOG(ERROR) << "raw row fetch: " << curr_storage_rows_to_fetch
+  //            << " start: " << curr_storage_start_row << " end: " <<
+  //            curr_storage_end_row;
+  if (curr_storage_rows_to_fetch <= curr_storage_start_row) {
+    curr_storage_start_row -= curr_storage_rows_to_fetch;
+  } else {
+    size_t to_fetch = std::min(curr_storage_rows_to_fetch - curr_storage_start_row,
+                               curr_storage_end_row - curr_storage_start_row);
+    // LOG(ERROR) << "emplacing from: " << curr_storage_start_row
+    //            << " to fetch: " << to_fetch;
+    retval.emplace_back(storage_->getUnderlyingBuffer() +
+                            storage_->getColOffInBytes(column_idx) +
+                            colType(column_idx)->size() * curr_storage_start_row,
+                        to_fetch);
+    curr_storage_end_row -= to_fetch;
+    total_handled_rows += to_fetch;
+    curr_storage_start_row = 0;
+  }
 
-  for (auto& chunk_uptr : appended_storage_) {
+  for (auto& storage_uptr : appended_storage_) {
     const int8_t* ptr =
-        chunk_uptr->getUnderlyingBuffer() + chunk_uptr->getColOffInBytes(column_idx);
-    size_t row_count = chunk_uptr->binSearchRowCount();
-    retval.emplace_back(ptr, row_count);
+        storage_uptr->getUnderlyingBuffer() + storage_uptr->getColOffInBytes(column_idx);
+    size_t row_count = storage_uptr->binSearchRowCount();
+    curr_storage_rows_to_fetch = row_count;
+    if (curr_storage_end_row == 0)
```

Comment:
`curr_storage_end_row == 0` means we have to stop, right? If we got all the required rows from the first storage, then `curr_storage_end_row` should be 0 here and we should stop.

Overall, it's hard to read the code because of misleading variable names. `curr_storage_rows_to_fetch` doesn't hold the number of rows to fetch. `curr_storage_start_row` and `curr_storage_end_row` sound like iterators to something but in fact they are not. `expected_row_count` is named and used as if it holds a total number of rows to fetch, but it holds the total number of rows in the result set instead and `(total_handled_rows >= expected_row_count)` check is useless.

Could you please just use `rows_to_fetch` and `rows_to_skip` variables instead of the 5 variables you currently have? Initialize them like
```
size_t rows_to_skip = getOffset();
size_t rows_to_fetch = getLimit() ? std::min(getLimit(), rowCount() - getOffset()) : rowCount() - getOffset();
```
Then use them to compute a number of rows to skip/fetch from each storage and adjust them accordingly, stopping when `rows_to_fetch` is zero. I believe that would make the code more readable.

Also, please pay attention to your tests. Try to trigger current errors in tests before fixing them.

Reply:
done